apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java

```diff
@@ -195,6 +197,153 @@ public Bundle searchByLogicalId(
     return bundle;
   }
 
+  @Search
+  public Bundle searchByCoverageContract(
+      // This is very explicit as a place holder until this kind
+      // of relational search is more common.
+      @RequiredParam(name = "_has:Coverage.extension") TokenParam coverageId,
+      @OptionalParam(name = "startIndex") String startIndex,
+      RequestDetails requestDetails) {
+
+    if (coverageId.getQueryParameterQualifier() != null)
+      throw new InvalidRequestException(
+          "Unsupported query parameter qualifier: " + coverageId.getQueryParameterQualifier());
+
+    CcwCodebookVariable partDContractMonth =
+        partDCwVariableFor(
+            coverageId.getSystem().substring(coverageId.getSystem().lastIndexOf('/') + 1));
+    SingularAttribute<Beneficiary, String> contractMonth = partDFieldFor(partDContractMonth);
+
+    String contractCode = coverageId.getValueNotNull();
+    if (contractCode.length() != 5)
+      throw new InvalidRequestException("Unsupported query parameter value: " + contractCode);
+
+    List<String> includeIdentifiersValues = returnIncludeIdentifiersValues(requestDetails);
+    List<SetAttribute<Beneficiary, ?>> withRelations =
+        new LinkedList<SetAttribute<Beneficiary, ?>>();
+    if (hasHICN(includeIdentifiersValues))
+      withRelations.add((SetAttribute<Beneficiary, ?>) Beneficiary_.beneficiaryHistories);
+
+    if (hasMBI(includeIdentifiersValues))
+      withRelations.add((SetAttribute<Beneficiary, ?>) Beneficiary_.medicareBeneficiaryIdHistories);
+
+    List<Beneficiary> matchingBeneficiaries =
+        queryBeneficiariesBy(contractMonth, contractCode, withRelations);
+
+    List<IBaseResource> patients =
+        matchingBeneficiaries.stream()
+            .map(
+                beneficiary -> {
+                  // Null out the unhashed HICNs if we're not supposed to be returning them
+                  if (!hasHICN(includeIdentifiersValues)) {
+                    beneficiary.setHicnUnhashed(Optional.empty());
+                  }
+                  // Null out the unhashed MBIs if we're not supposed to be returning
+                  if (!hasMBI(includeIdentifiersValues)) {
+                    beneficiary.setMedicareBeneficiaryId(Optional.empty());
+                  }
+
+                  Patient patient =
+                      BeneficiaryTransformer.transform(
+                          metricRegistry, beneficiary, includeIdentifiersValues);
+                  return patient;
+                })
+            .collect(Collectors.toList());
+
+    PagingArguments pagingArgs = new PagingArguments(requestDetails);
+    Bundle bundle =
+        TransformerUtils.createBundle(
+            pagingArgs,
+            "/Patient?",
+            "_has:Coverage.extension",
+            coverageId.getValueAsQueryToken(null),
+            patients);
+    return bundle;
```

Comment:
If my reading is correct, paging happens _post-query_, i.e. after all matching benes have been pulled from the DB. Can this become a problem though? Some of our Part D contracts have a lot of benes.

Reply:
Bene counts for some of the contracts we have:
![image](https://user-images.githubusercontent.com/11096575/73479267-ad82a680-4365-11ea-9e09-3aaf3baf7664.png)

Reply:
+1000 Yes, the paging should absolutely be done by the database. I'll start an attempt at moving them over.

If it turns out to require a more fundamental change than makes sense for this PR then I'll leave this as is and make sure designing the move is in the backlog. For what it's worth the post-db paging is used for all resources (including EoBs) and does work with large data sets.

Reply:
done.